alternator/scoped_item_list.hh

```diff
@@ -0,0 +1,181 @@
+    list_type _list;
+
+public:
+    template <typename... Args>
+    handle emplace(Args&&... args) {
+        handle ret(this, std::forward<Args>(args)...);
+        _list.push_back(ret);
+        return ret;
+    }
+    scoped_item_list() = default;
+    ~scoped_item_list() {
+        // We usually expect that when the scoped_item_list<T> is destroyed,
+        // the list is already empty. For example, in the Alternator server,
+        // the list is destroyed when the server object is destroyed, and the
+        // server is only destroyed after all requests are done - so all
+        // handles those requests held to list items have been destroyed, so
+        // the list is empty.
+        // But if for some reason the list is not empty, it's not a disaster
+        // either - we just remove the "owner" pointer from all the handles.
+        // Each handle will continue to hold its item as before, but it is
+        // detached from any list.
+        for (auto& h : _list) {
+            h.owner = nullptr;
+        }
+        // We don't expect that the scoped_item_list<T> be destroyed while any
+        // for_each_gently() are still running. In general trying to do that
+        // doesn't make sense - since for_each_gently() is a method of the
+        // object being destroyed. We could have allowed this by making the
+        // _gentle_iterators list an intrusive list as well, but it's not
+        // worth the effort, since we don't expect this to ever happen.
+        if (!_gentle_iterators.empty()) {
+            utils::on_fatal_internal_error("scoped_item_list destroyed while "
+                "for_each_gently() is still running");
+        }
+    }
+    scoped_item_list(const scoped_item_list&) = delete;
+    scoped_item_list& operator=(const scoped_item_list&) = delete;
+
+    bool empty() const { return _list.empty(); }
+
+    struct gentle_iterator {
+        list_type::iterator iter, end;
+        gentle_iterator(scoped_item_list<T>& s) : iter(s._list.begin()), end(s._list.end()) {}
+        gentle_iterator(const gentle_iterator&) = delete;
+        gentle_iterator(gentle_iterator&&) = delete;
+    };
+    std::list<gentle_iterator> _gentle_iterators;
+
+    // for_each_gently() is a stall-safe way to iterate over the list, i.e.,
+    // run a given function fn on each item in the list, possibly preempting
+    // between items when our time quota is finished, to avoid stalls.
+    // Note that the function fn cannot be preempted (it does not return a
+    // future) - it would not be safe if it did, and the item it was working
+    // on was destroyed.
+    future<> for_each_gently(noncopyable_function<void(T&)> fn) {
+        _gentle_iterators.emplace_front(*this);
+        typename std::list<gentle_iterator>::iterator gi = _gentle_iterators.begin();
+        return seastar::do_until([ gi ] { return gi->iter == gi->end; },
```

Comment:
I think it's okay to write this as a coroutine, accepting that the list is intended to be large, and for_each_gently is intended to be slow.

Reply:
I copied this for_each_gently code verbatim from generic_server.cc. Although I do understand what it does, I didn't want to rewrite it without reason, given it's already short and the do_until() is not what makes it complicated, why change it?